### PR TITLE
Return temp path for targetDir() of a single file project

### LINKED ISSUE #33660: targetDir() method returns a wrong path for single file projects
**Description:**
Please refer https://github.com/ballerina-platform/ballerina-lang/blob/stage-slbeta4/compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java#L86

We will have to check and handle this implementation for single file projects.
Currently it returns `singlefile.bal/target` as the target dir path.

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by Ibaqu**
Fixes #33660 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java [hunk 1]
```diff
@@ -82,6 +84,13 @@ public Path sourceRoot() {
     }
 
     public Path targetDir() {
+        if (this.projectKind == ProjectKind.SINGLE_FILE_PROJECT) {
+            try {
+                return Files.createTempDirectory("ballerina-cache" + System.nanoTime());
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java:89]: This will create temp directories each time we call the function. We can improve this as follows: 
1. Make `Project.target` abstract and implement it in SingleFileProject and BuildProjects. BalaProject should throw an unsupported operation exception.
2. In SingleFileProject, declare a class level variable (e.g. targetDir) and create and assign the temp location at the project creation
3. BuildProject can contain the existing implementation

Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java:89]: After this change, shall we try to refactor the tasks in CLI to use this method?

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/SingleFileProject.java [hunk 2]
```diff
@@ -34,6 +36,8 @@
  */
 public class SingleFileProject extends Project {
 
+    Path targetDir;
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/SingleFileProject.java:39]: This can be private final

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/bala/BalaProject.java [hunk 3]
```diff
@@ -126,4 +126,9 @@ private boolean isFilePathInProject(Path filepath) {
         }
         return true;
     }
+
+    @Override
+    public Path targetDir() {
+        throw new UnsupportedOperationException("Target Dir is not supported for BalaProject");
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/bala/BalaProject.java:132]: I have second thoughts on this. Should we rather return `null` and add it in javadocs for the method? @sameerajayasoma @hevayo wdyt?

Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/bala/BalaProject.java:132]: We'll go ahead with throwing the exception.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/SingleFileProject.java (hunk 2 above)
Comment (Ibaqu) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/SingleFileProject.java]: I changed it to private Path targetDir. final might not work since the initialization is done inside a try block and we are ignoring the exception entirely. WDYT?

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/bala/BalaProject.java (hunk 3 above)
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/bala/BalaProject.java:132]: Shall we reword this to `target directory`

Comment (Ibaqu) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/bala/BalaProject.java]: Changed to "Target directory is not supported for Balaproject"

**Comment by azinneera** [comment]
Shall we add test cases for these? Or rather we can modify the existing tests in project-api-test to assert the target directory path.

File: project-api/project-api-test/src/test/java/io/ballerina/projects/test/TestSingleFileProject.java [hunk 4]
```diff
@@ -74,6 +74,21 @@ public void testLoadSingleFile() {
...
+        SingleFileProject project = null;
+        try {
+            project = TestUtils.loadSingleFileProject(projectPath);
+        } catch (Exception e) {
+            Assert.fail(e.getMessage());
+        }
+
+        Path targetDirPath = project.targetDir();
+        Assert.assertNotNull(targetDirPath);
+        Assert.assertTrue(targetDirPath.toFile().exists());
```
Comment (azinneera) [inline project-api/project-api-test/src/test/java/io/ballerina/projects/test/TestSingleFileProject.java:89]: Lets add an assertion to make sure its a temp location

Comment (Ibaqu) [reply project-api/project-api-test/src/test/java/io/ballerina/projects/test/TestSingleFileProject.java:89]: Added extra assertion : `Assert.assertEquals(Path.of(System.getProperty("java.io.tmpdir")), targetDirPath.getParent());`

Comment (Ibaqu) [reply project-api/project-api-test/src/test/java/io/ballerina/projects/test/TestSingleFileProject.java:89]: Changed to us `Paths.get()`

**Review by azinneera — APPROVED** [review]
